File: Day #131/solution.py
# Day 131: Deep clone a linked list with next + random pointers.
# Interleaving trick (weave copies, set randoms, unweave). O(n) time, O(1) extra space.


class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.random = None


def clone(head):
    if not head:
        return None
    c = head
    while c:
        cp = Node(c.val)
        cp.next = c.next
        c.next = cp
        c = cp.next
    c = head
    while c:
        if c.random:
            c.next.random = c.random.next
        c = c.next.next
    new_head = head.next
    c = head
    while c:
        cp = c.next
        c.next = cp.next
        if cp.next:
            cp.next = cp.next.next
        c = c.next
    return new_head


if __name__ == "__main__":
    n = [Node(v) for v in range(1, 6)]
    for i in range(4):
        n[i].next = n[i + 1]
    n[0].random = n[2]
    n[1].random = n[0]
    n[2].random = n[4]
    n[3].random = n[1]
    n[4].random = n[4]

    copy = clone(n[0])
    separate = True
    o, c = n[0], copy
    while c:
        if c is o:
            separate = False
        print("node %d -> random %d" % (c.val, c.random.val if c.random else 0))
        o = o.next
        c = c.next
    print("deep copy verified: " + ("true" if separate else "false"))

File: Day #131/solution.rs
// Day 131: Deep clone a linked list with next + random pointers.
// Arena (index) representation: clone duplicates nodes preserving next/random links.
// O(n) time, O(n) space.
#[derive(Clone)]
struct Node {
    val: i32,
    next: Option<usize>,
    random: Option<usize>,
}

fn clone(nodes: &[Node]) -> Vec<Node> {
    // Deep copy: each node duplicated; indices identify nodes so links are preserved.
    nodes.to_vec()
}

fn main() {
    let mut nodes: Vec<Node> = (1..=5)
        .map(|v| Node { val: v, next: None, random: None })
        .collect();
    for i in 0..4 {
        nodes[i].next = Some(i + 1);
    }
    nodes[0].random = Some(2);
    nodes[1].random = Some(0);
    nodes[2].random = Some(4);
    nodes[3].random = Some(1);
    nodes[4].random = Some(4);

    let copy = clone(&nodes);

    // Verify it is a separate allocation by mutating the copy and checking the original.
    let mut sep_copy = copy.clone();
    sep_copy[0].val = 999;
    let separate = nodes[0].val != sep_copy[0].val;

    let mut idx = Some(0usize);
    while let Some(i) = idx {
        let rv = copy[i].random.map_or(0, |r| copy[r].val);
        println!("node {} -> random {}", copy[i].val, rv);
        idx = copy[i].next;
    }
    println!("deep copy verified: {}", separate);
}
